pkg/querier/handler/handler.go

```diff
@@ -0,0 +1,604 @@
+	// be called by the caller) or call qry.Close ourselves (which is
+	// required in the case of a panic).
+	defer func() {
+		if result.Finalizer == nil {
+			qry.Close()
+		}
+	}()
+
+	ctx = httputil.ContextFromRequest(ctx, r)
+
+	res := qry.Exec(ctx)
+	if res.Err != nil {
+		return ApiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
+	}
+
+	// Optional stats field in response if parameter "stats" is not empty.
+	sr := api.StatsRenderer
+	if sr == nil {
+		sr = defaultStatsRenderer
+	}
+	qs := sr(ctx, qry.Stats(), r.FormValue("stats"))
+
+	return ApiFuncResult{createPrometheusResponse(&queryData{
+		ResultType: res.Value.Type(),
+		Result:     res.Value,
+		Stats:      qs,
+	}), nil, res.Warnings, qry.Close}
+}
+
+func returnAPIError(err error) *apiError {
+	if err == nil {
+		return nil
+	}
+
+	cause := errors.Unwrap(err)
+	if cause == nil {
+		cause = err
+	}
+
+	switch cause.(type) {
+	case promql.ErrQueryCanceled:
+		return &apiError{errorCanceled, err}
+	case promql.ErrQueryTimeout:
+		return &apiError{errorTimeout, err}
+	case promql.ErrStorage:
+		return &apiError{errorInternal, err}
+	}
+
+	if errors.Is(err, context.Canceled) {
+		return &apiError{errorCanceled, err}
+	}
+
+	return &apiError{errorExec, err}
+}
+
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC()
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC()
+
+	minTimeFormatted = minTime.Format(time.RFC3339Nano)
+	maxTimeFormatted = maxTime.Format(time.RFC3339Nano)
+)
+
+func (api *API) Respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+	statusMessage := statusSuccess
+	var warningStrings []string
+	for _, warning := range warnings {
+		warningStrings = append(warningStrings, warning.Error())
+	}
+	var b []byte
+	var err error
+	switch data.(type) {
+	case queryrange.PrometheusResponse:
+		prometheusResponse, _ := data.(queryrange.PrometheusResponse)
+		prometheusResponse.Status = string(statusMessage)
+		b, err = proto.Marshal(&prometheusResponse)
+	case instantquery.PrometheusInstantQueryResponse:
+		prometheusInstantQueryResponse, _ := data.(instantquery.PrometheusInstantQueryResponse)
+		prometheusInstantQueryResponse.Status = string(statusMessage)
+		b, err = proto.Marshal(&prometheusInstantQueryResponse)
+	default:
+		level.Error(api.Logger).Log("msg", "error asserting response type")
+		http.Error(w, "error asserting response type", http.StatusInternalServerError)
+		return
+	}
+	if err != nil {
+		level.Error(api.Logger).Log("msg", "error marshaling protobuf response", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/protobuf")
+	w.WriteHeader(http.StatusOK)
+	if n, err := w.Write(b); err != nil {
+		level.Error(api.Logger).Log("msg", "error writing response", "bytesWritten", n, "err", err)
+	}
+}
+
+func (api *API) RespondError(w http.ResponseWriter, apiErr *apiError, data interface{}) {
```

Comment:
Do we still return error in json format? Query frontend requests protobuf format only so I guess we need to return protobuf format as well?

Reply:
Errors are actually intercepted in the Codec `DecodeResponse()` function and converted to a json directly skipping the protobuf part